File: go/problems/apply_discount_every_n_orders.go
package problems

type Cashier struct {
	threshold int
	ordersProcessed int
	discount int
	products []int
	prices []int
}


func Constructor(n int, discount int, products []int, prices []int) Cashier {
	return Cashier{
		threshold: n,
		discount: discount,
		products: products,
		prices: prices,
		ordersProcessed: 1,
	}
}


func (c *Cashier) GetBill(product []int, amount []int) float64 {
	if len(product) != len(amount) {
		panic("Invalid args!")
	}
	productIndicesMap := make(map[int]int)
	bill := 0.0
	for index, product := range c.products {
		productIndicesMap[product] = index
	}
	for index, prod := range product {
		productIndex := productIndicesMap[prod]
		bill += float64(c.prices[productIndex] * amount[index])
	}
	if c.ordersProcessed % c.threshold == 0 {
		bill -= (float64(c.discount) * bill) / 100.0
	}
	c.ordersProcessed += 1
	return bill
}

File: rust/src/problems/apply_discount_every_n_orders.rs
// https://leetcode.com/problems/apply-discount-every-n-orders/
pub trait Employee {
    fn new(n: i32, discount: i32, products: Vec<i32>, prices: Vec<i32>) -> Self;
    fn get_bill(&mut self, product: Vec<i32>, amount: Vec<i32>) -> f64;
}

pub struct Cashier {
    threshold: i32,
    orders_processed: i32,
    discount: i32,
    products: Vec<i32>,
    prices: Vec<i32>,
}

impl Employee for Cashier {
    fn new(n: i32, discount: i32, products: Vec<i32>, prices: Vec<i32>) -> Self {
        return Cashier {
            threshold: n,
            orders_processed: 1,
            discount,
            products,
            prices,
        };
    }

    fn get_bill(&mut self, product: Vec<i32>, amount: Vec<i32>) -> f64 {
        if product.len() != amount.len() {
            panic!("Invalid args!");
        }
        let mut bill_amount: f64 = product
            .into_iter()
            .enumerate()
            .map(|(_, prod)| {
                return self.products
                    .clone()
                    .into_iter()
                    .position(|p| { p == prod })
                    .unwrap();
            })
            .enumerate()
            .fold(0.0, |mut acc, (index, prod)| {
                let price = self.prices[prod];
                let quantity = amount[index];
                acc += (quantity * price) as f64;
                return acc;
            });
        if self.orders_processed % self.threshold == 0 {
            bill_amount -= (self.discount as f64 * bill_amount) / 100.0;
        }
        self.orders_processed += 1;
        return bill_amount;
    }
}